src/vs/workbench/api/browser/mainThreadCustomEditors.ts

```diff
@@ -106,6 +106,23 @@ export class MainThreadCustomEditors extends Disposable implements extHostProtoc
 
 		// Working copy operations
 		this._register(workingCopyFileService.onWillRunWorkingCopyFileOperation(async e => this.onWillRunWorkingCopyFileOperation(e)));
+
+		this._register(extensionService.onWillStop(e => {
+			const dirtyCustomEditors = workingCopyService.workingCopies.filter(workingCopy => {
+				return workingCopy instanceof MainThreadCustomEditorModel && workingCopy.isDirty();
+			});
+			if (!dirtyCustomEditors.length) {
+				return;
+			}
+
+			e.veto((async () => {
+				let didSaveAll = true;
+				for (const dirtyCustomEditor of dirtyCustomEditors) {
+					didSaveAll &&= await dirtyCustomEditor.save();
```

Comment:
Can this trigger a file dialog for untitled custom editors? If so, we should probably test that it `awaits` properly.

Reply:
It does await properly but using `&&=` is incorrect because we don't want short circuiting. Will push a fix for that